envoy/network/listener_filter_buffer.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * Interface for ListenerFilterBuffer
+ */
+class ListenerFilterBuffer {
+public:
+  virtual ~ListenerFilterBuffer() = default;
+  /**
+   * Copy the bufferred data into the address which `buffer` pointed to.
+   * But it won't drain the data after copyOut, except an explicit drain method invoked.
+   * @param buffer supplies the buffer to read into.
+   * @param max_length supplies the maximum length to read.
+   * @return the length of data read into the buffer.
+   */
+  virtual uint64_t copyOut(void* buffer, uint64_t length) PURE;
```

Comment:
I would have buffer be typed here, or just pass in a `Buffer::Instance` which I think will be safer.

Reply:
that will need to convert all the filter to use the Buffer::Instance. Currently, all the filter is based on raw buffer. Although I can turn them to use buffer, but there won't be any benefit from it. Anyway let me give a try, then see what it is looks like.